geopandas/geodataframe.py

```diff
@@ -1459,7 +1467,15 @@ def _constructor(self):
 
     @property
     def _constructor_sliced(self):
-        return _geoseries_constructor_with_fallback
+        def _geodataframe_constructor_sliced(data=None, index=None, crs=None, **kwargs):
+
+            srs = pd.Series(data, index, **kwargs)
+            is_col_proxy = srs.index is self.index
+            if is_geometry_type(srs) and is_col_proxy:
```

Comment:
One small thought: should we maybe invert the check? (not returning a GeoSeries if we are sure it is a row, instead of only returning a GeoSeries if we are sure it is a column, so something like `is_row_proxy = srs.index is self.columns` and then `.. and not is_row_proxy`) 
When selecting a column, there could maybe some reason that the index is not _identical_, and with this inverted check that would still be covered

Reply:
I think that makes sense given we're specialising the row case the check should be on the row case (it looks a bit weird in the code though because I upcast after the is geomery_type check, but I think that's just perspective)